perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_def_login_namespace_lifescience_hostel_persistent_shadow.java

```diff
@@ -0,0 +1,137 @@
+	 * @param attribute Attribute to fill value to
+	 * @return Filled attribute
+	 * @throws InternalErrorException
+	 * @throws WrongAttributeAssignmentException
+	 */
+	@Override
+	public Attribute fillAttribute(PerunSessionImpl perunSession, User user, AttributeDefinition attribute) throws InternalErrorException, WrongAttributeAssignmentException {
+
+		Attribute filledAttribute = new Attribute(attribute);
+
+		if (generatedNamespaces.contains(attribute.getFriendlyNameParameter())) {
+
+			ModulesUtilsBlImpl.LoginGenerator generator = new ModulesUtilsBlImpl.LoginGenerator();
+			String login = generator.generateLogin(user, new ModulesUtilsBlImpl.LoginGenerator.LoginGeneratorFunction() {
+				@Override
+				public String generateLogin(String firstName, String lastName) {
+					String login = firstName+ "." + lastName;
+					if (login.length()>20) {
+						login = login.substring(0, 20);
+					}
+					return login;
+				}
+			});
+
+			if (login == null) return filledAttribute;
+
+			// fill value
+			int iterator = 0;
+			while (iterator >= 0) {
+				if (iterator > 0) {
+					int iteratorLength = String.valueOf(iterator).length();
+					if (login.length() + iteratorLength > 20) {
+						// if login+iterator > 20 => crop login & reset iterator
+						login = login.substring(0, login.length()-1);
+						iterator = 0;
+						filledAttribute.setValue(login);
+					} else {
+						filledAttribute.setValue(login + iterator);
+					}
+
+				} else {
+					filledAttribute.setValue(login);
+				}
+				try {
+					checkAttributeValue(perunSession, user, filledAttribute);
+					return filledAttribute;
+				} catch (WrongAttributeValueException ex) {
+					// continue in a WHILE cycle
+					iterator++;
+				}
+			}
+
+			return filledAttribute;
+
+		} else {
+			// without value
+			return filledAttribute;
+		}
+	}
+
+	/**
+	 * ChangedAttributeHook() sets UserExtSource with following properties:
+	 *  - extSourceType is IdP
+	 *  - extSourceName is https://login.bbmri-eric.eu/idp/
+	 *  - user's extSource login is the same as his lifescience-hostel-persistent attribute
+	 *
+	 * @param session PerunSession
+	 * @param user User to set UserExtSource for
+	 * @param attribute Attribute containing bbmriID
+	 * @throws cz.metacentrum.perun.core.api.exceptions.InternalErrorException
+	 * @throws cz.metacentrum.perun.core.api.exceptions.WrongReferenceAttributeValueException
+	 */
+	@Override
+	public void changedAttributeHook(PerunSessionImpl session, User user, Attribute attribute) throws InternalErrorException, WrongReferenceAttributeValueException {
+		try {
+			String userNamespace = attribute.getFriendlyNameParameter();
+
+			if(userNamespace.equals("lifescience-hostel-persistent-shadow") && attribute.getValue() != null){
+				ExtSource extSource = session.getPerunBl().getExtSourcesManagerBl().getExtSourceByName(session, extSourceNameLSH);
+				UserExtSource userExtSource = new UserExtSource(extSource, 0, attribute.getValue().toString());
```

Comment:
We usually change type of resulting attribute value to String ``((String)attribute.getValue())`` instead of calling ``toString()`` on object.